std/file.d

```diff
@@ -3200,6 +3200,219 @@ version(Posix) unittest //issue 11434
     assert(getAttributes(t2) == octal!100767);
 }
 
+/***************************************************
+Copy file or directory $(D from) to $(D to). When a directory is copied, its
+contents are copied as well, recursively.
+
+File timestamps are preserved, except for symbolic links.
+File attributes are preserved, if $(D preserve) equals $(D PreserveAttributes.yes).
+On Windows only $(D PreserveAttributes.yes) (the default on Windows) is supported.
+
+Existing target files are overwritten. Existing target directories are written
+into.
+
+With symbolic links, the links themselves are copied. That is, for every
+symbolic link in $(D from), a new link with the same target path is created in
+$(D to). Relative target paths are not changed. They may be invalidated by the
+copying, when they point outside of the copied directory tree.
+
+Special files (e.g. device files, FIFOs) cannot be copied with $(D copyRecurse).
+
+Params:
+    from = string or range of characters representing the existing
+        file/directory name
+    to = string or range of characters representing the target file/directory
+        name
+    preserve = whether to preserve the file attributes
+
+Throws: $(D FileException) on error.
+ */
+void copyRecurse(string from, string to,
+    PreserveAttributes preserve = preserveAttributesDefault)
+{
+    // Remove any trailing directory separators.
+    from = buildNormalizedPath(from);
+    to = buildNormalizedPath(to);
+
+    immutable attrs = getLinkAttributes(from);
+
+    if (attrIsFile(attrs)) copy(from, to, preserve);
+    else if (attrIsDir(attrs))
+    {
+        mkdirRecurse(std.conv.to!string(to));
+        
+        if (preserve == PreserveAttributes.yes) setAttributes(to, attrs);
+        
+        foreach (entry; dirEntries(from, SpanMode.shallow, false))
+        {
+            immutable path = entry.name[from.length + 1 .. $];
+                // + 1 for the directory separator
+            copyRecurse(entry.name, buildPath(to, path), preserve);
+        }
+        
+        SysTime accessTime, modificationTime;
+        getTimes(from, accessTime, modificationTime);
+        setTimes(to, accessTime, modificationTime);
+    }
+    else if (attrIsSymlink(attrs))
+    {
+        version (Posix)
+        {
+            if (exists(to)) remove(to);
+            symlink(readLink(from), to);
+            if (preserve == PreserveAttributes.yes)
+            {
+                /*NOTE: Would copy attributes of the symlink here, but there
+                is no setLinkAttributes here, and no fchmodat in the C headers.
+                Just enforcing that the new attributes are the same as the old
+                ones, for now. */
```

Comment:
Fixing it would mean that a call to copyRecurse then properly copies everything instead of throwing an exception. Depending on that particular exception would seem silly to me. But, yeah, I guess it's technically a breaking change.

We could settle with the throwing behaviour for good, of course. Or maybe it's possible to implement copying of symlink attributes. I'm not so well informed on Linux/Posix/Unix stuff, but as far as I could find out, fchmodat is needed to do it. `man 2 fchmodat` says that it "was added to Linux in kernel 2.6.16; library support was added to glibc in version 2.4.", and that it's in POSIX.1-2008. That seems rather new. I don't know if it's too new for us to use.

Also note that most systems apparently don't even have a way to change the permissions of symlinks. OS X seems to be the outlier in that it does not have fixed symlink permissions.